### PR TITLE
[libc][math][c23] Add fmodl and fmodf128 math functions.

### PR DESCRIPTION
**Pull request by lntue**
- Allow `FMod` template to have different computational types and make it work for 80-bit long double.
- Switch to use `uint64_t` as the intermediate computational types for `float`, significantly reduce the latency of `fmodf` when the exponent difference is large.

### CODE REVIEW

File: libc/src/__support/FPUtil/generic/FMod.h [hunk 1]
```diff
@@ -219,22 +158,57 @@ template <typename T> class FModDivisionInvMultHelper {
...
   using StorageType = typename FPB::StorageType;
 
+  LIBC_INLINE static bool pre_check(T x, T y, T &out) {
+    using FPB = fputil::FPBits<T>;
+    const T quiet_nan = FPB::quiet_nan().get_val();
+    FPB sx(x), sy(y);
+    if (LIBC_LIKELY(!sy.is_zero() && !sy.is_inf_or_nan() &&
+                    !sx.is_inf_or_nan())) {
+      return false;
+    }
```
Comment (nickdesaulniers) [inline libc/src/__support/FPUtil/generic/FMod.h:179]: remove braces for single statement conditionals

File: libc/src/__support/FPUtil/generic/FMod.h [hunk 2]
```diff
@@ -219,22 +158,57 @@ template <typename T> class FModDivisionInvMultHelper {
...
+
+    if (sy.is_inf()) {
+      out = x;
+      return true;
+    }
+
+    // case where x == 0
+    out = x;
+    return true;
```
Comment (nickdesaulniers) [inline libc/src/__support/FPUtil/generic/FMod.h:203]: Either sides of this branch look the same to me. Intentional?

File: libc/src/__support/FPUtil/generic/FMod.h [hunk 3]
```diff
@@ -219,22 +158,57 @@ template <typename T> class FModDivisionInvMultHelper {
...
+    const T quiet_nan = FPB::quiet_nan().get_val();
+    FPB sx(x), sy(y);
+    if (LIBC_LIKELY(!sy.is_zero() && !sy.is_inf_or_nan() &&
+                    !sx.is_inf_or_nan())) {
+      return false;
+    }
+
+    if (sx.is_nan() || sy.is_nan()) {
+      if ((sx.is_nan() && !sx.is_quiet_nan()) ||
+          (sy.is_nan() && !sy.is_quiet_nan()))
```
Comment (nickdesaulniers) [inline libc/src/__support/FPUtil/generic/FMod.h:183]: Why not simply:
```suggestion
   if ((sx.is_nan() && !sx.is_quiet_nan()) ||
       (sy.is_nan() && !sy.is_quiet_nan()))
```

Comment (lntue) [reply libc/src/__support/FPUtil/generic/FMod.h]: It's a bit subtle, that those 2 are not equivalent.  Line 181 means either of them is nan, and lines 182-183 means either of them is signalling nan.  I updated L182-183 to be a bit clearer.

File: libc/src/__support/FPUtil/generic/FMod.h [hunk 4]
```diff
@@ -174,13 +174,11 @@ class FMod {
...
     if (LIBC_LIKELY(!sy.is_zero() && !sy.is_inf_or_nan() &&
-                    !sx.is_inf_or_nan())) {
+                    !sx.is_inf_or_nan()))
       return false;
-    }
 
     if (sx.is_nan() || sy.is_nan()) {
-      if ((sx.is_nan() && !sx.is_quiet_nan()) ||
-          (sy.is_nan() && !sy.is_quiet_nan()))
+      if (sx.is_signalling_nan() || sy.is_signalling())
```
Comment (nickdesaulniers) [inline libc/src/__support/FPUtil/generic/FMod.h:181]: Should this be using either `is_signalling_nan` or `is_signalling` for BOTH `sx` AND `sy`?

Comment (lntue) [reply libc/src/__support/FPUtil/generic/FMod.h]: Actually neither of them are correct.  The correct function is `is_signaling_nan()`.

**Review by nickdesaulniers — APPROVED** [review]

**Pull request merged**